File: runtime/contracts/workspace/trusted-fill.js
import { randomBytes, randomUUID } from 'node:crypto';
import { exact } from './automation.js';
import { claimTime } from './claim-time.js';
import { copy, fromJSON, get, int, integer, object, same, set, string, text, JobsError } from './values.js';
export const trustedFillPolicyRevision = 1n;
export const trustedFillOperations = new Set([
    'fill_text', 'select_option', 'toggle_non_consent', 'upload_approved_resume',
]);
const fingerprints = ['urlFingerprint', 'observedQuestionFingerprint', 'observedControlFingerprint', 'formFingerprint'];
const revisions = ['jobRevision', 'resumeRevision', 'profileRevision', 'vitalFactRevision',
    'automationSettingsRevision', 'policyRevision', 'approvalRevision'];
const approvalFields = ['schemaVersion', 'approvalId', 'status', 'issuedAt', 'expiresAt', 'revokedAt',
    'jobId', 'jobRevision', 'claimId', 'realmRef', 'urlFingerprint', 'resumeId', 'resumeRevision',
    'resumeContentRevision', 'profileRevision', 'vitalFactRevision', 'answerBindings',
    'observedQuestionFingerprint', 'observedControlFingerprint', 'formFingerprint',
    'automationSettingsRevision', 'employerAccountRevision', 'policyRevision', 'allowedOperations',
    'nonce', 'approvalRevision'];
const reference = /^[a-z][a-z0-9._-]{0,127}$/;
const fingerprintPattern = /^sha256:[0-9a-f]{64}$/;
const contentRevisionPattern = /^content_[A-Za-z0-9_-]{32,128}$/;
function positive(value, label) {
    const result = int(value);
    if (result === null || result < 1n)
        throw new JobsError(`${label} must be a positive integer`);
    return result;
}
function time(value, label) {
    const raw = string(value);
    if (!raw)
        throw new JobsError(`${label} is invalid`);
    try {
        return claimTime(raw);
    }
    catch {
        throw new JobsError(`${label} is invalid`);
    }
}
function fingerprint(value, label) {
    const result = string(value);
    if (!result || !fingerprintPattern.test(result))
        throw new JobsError(`${label} must be a sha256 fingerprint`);
    return result;
}
export function trustedFillContentRevision(value) {
    const result = string(value);
    if (!result || !contentRevisionPattern.test(result))
        throw new JobsError('resume content revision is unverifiable');
    return result;
}
export function trustedFillOperationList(value, label = 'allowed operations') {
    if (!Array.isArray(value) || !value.length || value.some(item => string(item) === null)) {
        throw new JobsError(`${label} are invalid or include a final action`);
    }
    const result = value.map(item => string(item));
    if (new Set(result).size !== result.length || result.some(item => !trustedFillOperations.has(item))) {
        throw new JobsError(`${label} are invalid or include a final action`);
    }
    return result.sort();
}
export function validateTrustedFillApproval(value) {
    const approval = object(value, 'trusted fill approval');
    exact(approval, approvalFields, 'approval contains unsupported fields');
    if (int(get(approval, 'schemaVersion')) !== 1n)
        throw new JobsError('approval contains unsupported fields');
    for (const field of ['approvalId', 'jobId', 'resumeId']) {
        if (!reference.test(string(get(approval, field)) ?? ''))
            throw new JobsError('approval reference is invalid');
    }
    if (!/^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/.test(string(get(approval, 'claimId')) ?? '')) {
        throw new JobsError('approval claim binding is invalid');
    }
    const status = string(get(approval, 'status'));
    if (!['active', 'revoked'].includes(status ?? ''))
        throw new JobsError('approval status is invalid');
    const issued = time(get(approval, 'issuedAt'), 'approval issuedAt');
    const expires = time(get(approval, 'expiresAt'), 'approval expiresAt');
    if (expires <= issued || expires - issued > 3600000000n)
        throw new JobsError('approval duration is invalid');
    if (status === 'revoked')
        time(get(approval, 'revokedAt'), 'approval revokedAt');
    else if (get(approval, 'revokedAt') !== null)
        throw new JobsError('active approval cannot have revokedAt');
    if (!/^[0-9a-f]{64}$/.test(string(get(approval, 'realmRef')) ?? ''))
        throw new JobsError('approval realm reference is invalid');
    for (const field of fingerprints)
        fingerprint(get(approval, field), field);
    for (const field of revisions)
        positive(get(approval, field), field);
    if (int(get(approval, 'policyRevision')) !== trustedFillPolicyRevision)
        throw new JobsError('approval policy revision is unsupported');
    trustedFillContentRevision(get(approval, 'resumeContentRevision'));
    if (get(approval, 'employerAccountRevision') !== null)
        positive(get(approval, 'employerAccountRevision'), 'employerAccountRevision');
    const bindings = get(approval, 'answerBindings');
    if (!Array.isArray(bindings))
        throw new JobsError('answer bindings must be a list');
    const refs = bindings.map(value => {
        const binding = object(value, 'answer binding');
        exact(binding, ['answerRef', 'questionRevision', 'answerRevision'], 'answer binding is invalid');
        const answerRef = string(get(binding, 'answerRef'));
        if (!answerRef || !reference.test(answerRef))
            throw new JobsError('answer binding reference is invalid');
        positive(get(binding, 'questionRevision'), 'questionRevision');
        positive(get(binding, 'answerRevision'), 'answerRevision');
        return answerRef;
    });
    if (new Set(refs).size !== refs.length || refs.some((item, index) => index > 0 && refs[index - 1] > item)) {
        throw new JobsError('answer bindings must be unique and sorted');
    }
    trustedFillOperationList(get(approval, 'allowedOperations'));
    if (!/^[A-Za-z0-9_-]{32,128}$/.test(string(get(approval, 'nonce')) ?? ''))
        throw new JobsError('approval nonce is invalid');
    return approval;
}
export function validateTrustedFillDocument(value) {
    const document = object(value, 'trusted fill approvals');
    exact(document, ['schemaVersion', 'approvals', 'metadata'], 'trusted fill approval document contains unsupported fields');
    if (int(get(document, 'schemaVersion')) !== 1n)
        throw new JobsError('trusted fill approval schema version is unsupported');
    const approvals = object(get(document, 'approvals'), 'trusted fill approvals');
    for (const [jobId, value] of approvals.entries())
        if (string(jobId) !== string(get(validateTrustedFillApproval(value), 'jobId'))) {
            throw new JobsError('trusted fill approval job identity is invalid');
        }
    const metadata = object(get(document, 'metadata'), 'trusted fill metadata');
    exact(metadata, ['createdAt', 'updatedAt'], 'trusted fill metadata is invalid');
    for (const field of ['createdAt', 'updatedAt'])
        if (!string(get(metadata, field)))
            throw new JobsError('trusted fill metadata timestamp is invalid');
    return document;
}
export function trustedFillApproval(bindings, durationMinutes, approvalRevision, now) {
    if (durationMinutes < 1n || durationMinutes > 60n)
        throw new JobsError('approval duration must be between 1 and 60 minutes');
    const issued = claimTime(now), expires = issued + durationMinutes * 60000000n;
    const approval = copy(bindings);
    for (const [key, value] of Object.entries({ schemaVersion: 1, approvalId: `trusted-fill-${randomUUID()}`,
        status: 'active', issuedAt: now, expiresAt: new Date(Number(expires / 1000n)).toISOString().replace('.000Z', 'Z'),
        revokedAt: null, policyRevision: 1, nonce: randomBytes(32).toString('base64url') }))
        set(approval, key, fromJSON(value));
    set(approval, 'approvalRevision', integer(approvalRevision));
    return validateTrustedFillApproval(approval);
}
export function revokeTrustedFillApproval(value, expectedRevision, now) {
    const current = validateTrustedFillApproval(value);
    if (int(get(current, 'approvalRevision')) !== expectedRevision)
        throw new JobsError('approval revision conflict');
    if (string(get(current, 'status')) !== 'active')
        throw new JobsError('approval is not active');
    const updated = copy(current);
    set(updated, 'status', text('revoked'));
    set(updated, 'revokedAt', text(now));
    set(updated, 'approvalRevision', integer(expectedRevision + 1n));
    return validateTrustedFillApproval(updated);
}
export function publicTrustedFillStatus(value, now) {
    if (value === null)
        return object(fromJSON({ status: 'missing', approvalRevision: null }), 'trusted fill public status');
    const record = validateTrustedFillApproval(value), active = string(get(record, 'status')) === 'active';
    const result = object(fromJSON({ status: active && claimTime(now) >= time(get(record, 'expiresAt'), 'approval expiresAt') ? 'expired' : string(get(record, 'status')),
        jobId: string(get(record, 'jobId')), realmRef: string(get(record, 'realmRef')), expiresAt: string(get(record, 'expiresAt')),
        approvalRevision: null }), 'trusted fill public status');
    set(result, 'approvalRevision', get(record, 'approvalRevision'));
    return set(result, 'allowedOperations', trustedFillOperationList(get(record, 'allowedOperations')).map(text));
}
export function trustedFillDecision(record, current, observed, now) {
    if (string(get(record, 'status')) !== 'active')
        return object(fromJSON({ authorized: false, reasonCode: 'approval_revoked', retryAllowed: false }), 'decision');
    if (claimTime(now) >= time(get(record, 'expiresAt'), 'approval expiresAt'))
        return object(fromJSON({ authorized: false, reasonCode: 'approval_expired', retryAllowed: false }), 'decision');
    const flags = { authenticationRequired: 'authentication_required', consentRequired: 'consent_required',
        credentialFieldsPresent: 'credential_fields_present', finalControlsPresent: 'final_controls_present', unseenQuestions: 'unseen_questions', unseenControls: 'unseen_controls' };
    for (const [field, reason] of Object.entries(flags))
        if (get(observed, field) === true)
            return object(fromJSON({ authorized: false, reasonCode: reason, retryAllowed: false }), 'decision');
    const bound = ['jobId', 'jobRevision', 'claimId', 'realmRef', 'urlFingerprint', 'resumeId', 'resumeRevision',
        'resumeContentRevision', 'profileRevision', 'vitalFactRevision', 'answerBindings', 'automationSettingsRevision',
        'employerAccountRevision', 'policyRevision'];
    if (bound.some(field => !same(get(record, field), get(current, field))))
        return object(fromJSON({ authorized: false, reasonCode: 'canonical_drift', retryAllowed: false }), 'decision');
    if (fingerprints.slice(1).some(field => !same(get(record, field), get(observed, field))))
        return object(fromJSON({ authorized: false, reasonCode: 'observed_drift', retryAllowed: false }), 'decision');
    const operations = trustedFillOperationList(get(observed, 'fieldOperations'), 'field operations');
    const allowed = new Set(trustedFillOperationList(get(record, 'allowedOperations')));
    if (operations.some(operation => !allowed.has(operation)))
        return object(fromJSON({ authorized: false, reasonCode: 'operation_not_approved', retryAllowed: false }), 'decision');
    const result = object(fromJSON({ authorized: true, reasonCode: 'authorized_non_final_fields', retryAllowed: false,
        approvalRevision: null, allowedOperations: operations }), 'decision');
    return set(result, 'approvalRevision', get(record, 'approvalRevision'));
}

File: scripts/job_apply_store/domains/accounts/trusted_fill.py
"""Root-local trusted fill operations for Store composition."""

from __future__ import annotations

import copy
import hashlib
import sys
import uuid
from datetime import datetime
from typing import Any
from urllib.parse import urlsplit

from ... import accounts_runtime, constants, io, normalization
from ...errors import StoreError, TrustedFillCurrentError


_RUNTIME_PROVIDER = lambda: {}


def _bind_runtime(provider) -> None:
    """Bind this leaf to its owning facade's late-bound collaborators."""
    global _RUNTIME_PROVIDER
    _RUNTIME_PROVIDER = provider


def _late(name: str):
    runtime = _RUNTIME_PROVIDER()
    if name in runtime:
        return runtime[name]
    if name.endswith("_MODULE"):
        return accounts_runtime.companion({
            "ACCOUNTS_MODULE": "job_apply_accounts",
            "CANARY_EXECUTOR_MODULE": "job_apply_account_canary_executor",
            "ACCOUNT_FLOWS_MODULE": "job_apply_account_flows",
            "PASSWORD_ACCOUNT_FLOWS_MODULE": "job_apply_password_account_flows",
            "CREDENTIALS_MODULE": "job_apply_credentials",
            "TRUSTED_FILL_MODULE": "job_apply_trusted_fill",
        }[name])
    return _CANONICAL[name]


_CANONICAL = {
    'AGENT_BLOCKER_TYPE_BY_CODE': constants.AGENT_BLOCKER_TYPE_BY_CODE,
    'SCHEMA_VERSION': constants.SCHEMA_VERSION,
    '_require_object': io.require_object,
    '_safe_session_id': normalization._safe_session_id,
    'atomic_write_json': io.atomic_write_json,
    'copy': copy,
    'exclusive_file_lock': io.exclusive_file_lock,
    'hashlib': hashlib,
    'read_json_object': io.read_json_object,
    'uuid': uuid,
    'validate_version': io.validate_version,
}


class TrustedFillMixin:
    """Plain mixin; persistent state belongs to StoreBase."""

    def _load_trusted_fill_document(self) -> dict[str, Any]:
        document = _late('read_json_object')(self.trusted_fill_path, "trusted fill approvals")
        _late('validate_version')(document, "trusted fill approvals")
        if set(document) != {"schemaVersion", "approvals", "metadata"}:
            raise StoreError("trusted fill approval document contains unsupported fields")
        approvals = _late('_require_object')(document.get("approvals"), "trusted fill approvals")
        metadata = _late('_require_object')(document.get("metadata"), "trusted fill metadata")
        if set(metadata) != {"createdAt", "updatedAt"}:
            raise StoreError("trusted fill metadata is invalid")
        for field in ("createdAt", "updatedAt"):
            if not isinstance(metadata[field], str) or not metadata[field]:
                raise StoreError("trusted fill metadata timestamp is invalid")
        for job_id, approval in approvals.items():
            _late('_safe_session_id')(job_id)
            try:
                _late('TRUSTED_FILL_MODULE').validate_approval(approval)
            except _late('TRUSTED_FILL_MODULE').TrustedFillError as error:
                raise StoreError(str(error)) from None
            if approval["jobId"] != job_id:
                raise StoreError("trusted fill approval job identity is invalid")
        return document


    def _ensure_trusted_fill_document(self) -> None:
        with _late('exclusive_file_lock')(self.store_lock_path):
            if self.trusted_fill_path.exists():
                return
            now = self._now()
            _late('atomic_write_json')(self.trusted_fill_path, {
                "schemaVersion": _late('SCHEMA_VERSION'),
                "approvals": {},
                "metadata": {"createdAt": now, "updatedAt": now},
            })


    @staticmethod
    def _trusted_fill_fingerprint(value: str) -> str:
        return "sha256:" + _late('hashlib').sha256(value.encode("utf-8")).hexdigest()


    def _trusted_fill_current_locked(
        self, job_id: str, answer_refs: list[str]
    ) -> dict[str, Any]:
        jobs = self._load_jobs_document()["jobs"]
        job = jobs.get(job_id)
        if job is None or job.get("deletedAt") is not None or job["status"] != "in_progress":
            raise StoreError("trusted fill requires an in-progress claimed job")
        realm = _late('ACCOUNTS_MODULE').normalize_realm(job["url"])
        if realm["status"] != "resolved":
            raise StoreError("trusted fill portal realm is unresolved")
        try:
            preflight = self._preflight_job_record(job)
        except Exception:
            raise TrustedFillCurrentError("resume_observation_failed") from None
        if not preflight["ready"]:
            if "resume_file_missing" in preflight["errors"] or "resume_missing" in preflight["errors"]:
                reason = "resume_content_missing"
            elif "resume_file_changed" in preflight["errors"]:
                reason = "resume_content_changed"
            else:
                reason = "resume_preflight_not_ready"
            raise TrustedFillCurrentError(reason)
        if preflight["resumeId"] is None:
            raise TrustedFillCurrentError("resume_content_missing")
        resume = self._load_resumes_document()["resumes"].get(preflight["resumeId"])
        if resume is None or resume.get("deletedAt") is not None:
            raise TrustedFillCurrentError("resume_content_missing")
        if resume.get("storageKind") != "managed":
            raise TrustedFillCurrentError("resume_content_unverifiable")
        try:
            content_revision = _late('TRUSTED_FILL_MODULE').validate_content_revision(
                resume.get("contentRevision")
            )
        except _late('TRUSTED_FILL_MODULE').TrustedFillError:
            raise TrustedFillCurrentError("resume_content_unverifiable") from None
        profile_revision = self._load_profile_document()["metadata"].get("revision", 1)
        answers = self._load_answers_document()["answers"]
        answer_bindings = []
        for answer_ref in sorted(answer_refs):
            answer = answers.get(answer_ref)
            if (
                answer is None or answer.get("deletedAt") is not None
                or answer.get("reviewStatus", "accepted") != "accepted"
            ):
                raise TrustedFillCurrentError("answer_binding_invalid")
            answer_bindings.append({
                "answerRef": answer_ref,
                "questionRevision": answer["revision"],
                "answerRevision": answer["revision"],
            })
        settings = self._load_automation_settings_document()["settings"]
        account = self._load_employer_accounts_document()["accounts"].get(realm["realmRef"])
        return {
            "jobId": job_id,
            "jobRevision": job["revision"],
            "realmRef": realm["realmRef"],
            "urlFingerprint": self._trusted_fill_fingerprint(job["normalizedUrl"]),
            "resumeId": resume["id"],
            "resumeRevision": resume["revision"],
            "resumeContentRevision": content_revision,
            "profileRevision": profile_revision,
            "vitalFactRevision": profile_revision,
            "answerBindings": answer_bindings,
            "automationSettingsRevision": settings["revision"],
            "employerAccountRevision": account["revision"] if account is not None else None,
            "policyRevision": _late('TRUSTED_FILL_MODULE').POLICY_REVISION,
        }


    def approve_trusted_fill(self, incoming: dict[str, Any], *, public: bool = False) -> dict[str, Any]:
        packet = _late('_require_object')(incoming, "trusted fill approval request")
        required = {
            "jobId", "expectedJobRevision", "realmRef", "answerRefs",
            "observedQuestionFingerprint", "observedControlFingerprint",
            "formFingerprint", "allowedOperations", "durationMinutes",
        }
        if set(packet) != required:
            raise StoreError("trusted fill approval request contains unsupported fields")
        job_id = _late('_safe_session_id')(packet.get("jobId", ""))
        if not isinstance(packet.get("answerRefs"), list) or not all(isinstance(item, str) for item in packet["answerRefs"]):
            raise StoreError("trusted fill answer references must be a list of strings")
        if len(packet["answerRefs"]) != len(set(packet["answerRefs"])):
            raise StoreError("trusted fill answer references contain duplicates")
        self.initialize()
        self._ensure_account_control_documents()
        self._ensure_coordinator_files()
        self._ensure_trusted_fill_document()
        with _late('exclusive_file_lock')(self.store_lock_path):
            claim = self._load_coordinator_document()["claim"]
            if (
                claim is None or claim["jobId"] != job_id
                or self._now_datetime() >= self._parse_time(claim["expiresAt"])
            ):
                raise StoreError("trusted fill approval requires the live claimed job")
            try:
                current = self._trusted_fill_current_locked(job_id, packet["answerRefs"])
            except TrustedFillCurrentError as error:
                job = self._load_jobs_document()["jobs"][job_id]
                handed_off = self._trusted_fill_attention_handoff_locked(
                    job, error.reason_code
                )
                return {
                    "authorized": False,
                    "reasonCode": error.reason_code,
                    "retryAllowed": False,
                    "attentionHandoff": True,
                    "job": {
                        "id": handed_off["id"],
                        "status": handed_off["status"],
                        "revision": handed_off["revision"],
                    },
                }
            current["claimId"] = claim["claimId"]
            if current["jobRevision"] != packet.get("expectedJobRevision"):
                raise StoreError("job revision conflict")
            if current["realmRef"] != packet.get("realmRef"):
                raise StoreError("trusted fill realm binding mismatch")
            document = self._load_trusted_fill_document()
            previous = document["approvals"].get(job_id)
            if previous is not None and previous["status"] == "active" and self._now_datetime() < _late('TRUSTED_FILL_MODULE')._time(previous["expiresAt"], "approval expiresAt"):
                raise StoreError("active trusted fill approval already exists")
            bindings = {
                **current,
                "observedQuestionFingerprint": packet["observedQuestionFingerprint"],
                "observedControlFingerprint": packet["observedControlFingerprint"],
                "formFingerprint": packet["formFingerprint"],
                "allowedOperations": packet["allowedOperations"],
            }
            approval_revision = 1 if previous is None else previous["approvalRevision"] + 1
            try:
                approval = _late('TRUSTED_FILL_MODULE').create_approval(
                    bindings, packet["durationMinutes"], approval_revision, self._now_datetime()
                )
            except _late('TRUSTED_FILL_MODULE').TrustedFillError as error:
                raise StoreError(str(error)) from None
            document["approvals"][job_id] = approval
            document["metadata"]["updatedAt"] = self._now()
            _late('atomic_write_json')(self.trusted_fill_path, document)
        return _late('TRUSTED_FILL_MODULE').public_status(approval, self._now_datetime()) if public else approval


    def trusted_fill_status(self, job_id: str, *, public: bool = False) -> dict[str, Any] | None:
        self.initialize()
        self._ensure_trusted_fill_document()
        _late('_safe_session_id')(job_id)
        record = self._load_trusted_fill_document()["approvals"].get(job_id)
        if public:
            return _late('TRUSTED_FILL_MODULE').public_status(record, self._now_datetime())
        return _late('copy').deepcopy(record) if record is not None else None


    def revoke_trusted_fill(
        self, job_id: str, expected_approval_revision: int, *, public: bool = False
    ) -> dict[str, Any]:
        self.initialize()
        self._ensure_trusted_fill_document()
        _late('_safe_session_id')(job_id)
        with _late('exclusive_file_lock')(self.store_lock_path):
            document = self._load_trusted_fill_document()
            record = document["approvals"].get(job_id)
            if record is None:
                raise StoreError("trusted fill approval does not exist")
            try:
                updated = _late('TRUSTED_FILL_MODULE').revoke_approval(
                    record, expected_approval_revision, self._now_datetime()
                )
            except _late('TRUSTED_FILL_MODULE').TrustedFillError as error:
                raise StoreError(str(error)) from None
            document["approvals"][job_id] = updated
            document["metadata"]["updatedAt"] = self._now()
            _late('atomic_write_json')(self.trusted_fill_path, document)
        return _late('TRUSTED_FILL_MODULE').public_status(updated, self._now_datetime()) if public else updated


    def _trusted_fill_attention_handoff_locked(
        self, job: dict[str, Any], reason_code: str
    ) -> dict[str, Any]:
        claim = self._load_coordinator_document()["claim"]
        if (
            claim is None or claim["jobId"] != job["id"]
            or self._now_datetime() >= self._parse_time(claim["expiresAt"])
        ):
            raise StoreError("trusted fill denial requires the live claimed job")
        now = self._now()
        blocker_code = {
            "authentication_required": "login-required",
            "credential_fields_present": "login-required",
            "consent_required": "consent-required",
            "approval_missing": "owner-input-required",
            "approval_revoked": "owner-input-required",
            "approval_expired": "owner-input-required",
            "approval_revision_mismatch": "owner-input-required",
            "answer_binding_invalid": "owner-input-required",
            "resume_content_missing": "owner-input-required",
            "resume_content_unverifiable": "owner-input-required",
            "resume_observation_failed": "owner-input-required",
            "resume_preflight_not_ready": "owner-input-required",
            "unseen_questions": "owner-input-required",
        }.get(reason_code, "browser-state-uncertain")
        session = self._build_session(job["id"], {
            "status": "active",
            "step": f"trusted_fill_denied:{reason_code}",
            "answerKeys": [],
            "pendingFields": [],
            "attemptRevision": job["revision"],
            "blockers": [{
                "type": _late('AGENT_BLOCKER_TYPE_BY_CODE')[blocker_code],
                "code": blocker_code,
            }],
            "browserHandoff": {
                "state": "required", "reasonCode": blocker_code, "revision": 1,
            },
        }, now, expected_attempt_revision=job["revision"], expected_ats=job.get("ats"))
        operation_id = str(_late('uuid').uuid4())
        self._commit_coordinator_operation_locked({
            "kind": "handoff", "operationId": operation_id, "jobId": job["id"],
            "sourceStatus": "in_progress", "targetStatus": "needs_info",
            "expectedRevision": job["revision"], "at": now, "session": session,
            "historyEvent": self._history_event_for_operation(
                operation_id, job, "job-blocked", "needs_info", now
            ),
            "resultClaim": None,
        })
        return self._load_jobs_document()["jobs"][job["id"]]


    def evaluate_trusted_fill(self, incoming: dict[str, Any], *, public: bool = False) -> dict[str, Any]:
        observed = _late('_require_object')(incoming, "trusted fill evaluation")
        required = {
            "jobId", "expectedApprovalRevision", "observedQuestionFingerprint",
            "observedControlFingerprint", "formFingerprint", "fieldOperations",
            "authenticationRequired", "consentRequired", "credentialFieldsPresent",
            "finalControlsPresent", "unseenQuestions", "unseenControls",
        }
        if set(observed) != required:
            raise StoreError("trusted fill evaluation contains unsupported fields")
        for field in (
            "authenticationRequired", "consentRequired", "credentialFieldsPresent",
            "finalControlsPresent", "unseenQuestions", "unseenControls",
        ):
            if not isinstance(observed[field], bool):
                raise StoreError("trusted fill evaluation flags must be booleans")
        job_id = _late('_safe_session_id')(observed.get("jobId", ""))
        self.initialize()
        self._ensure_account_control_documents()
        self._ensure_coordinator_files()
        self._ensure_trusted_fill_document()
        with _late('exclusive_file_lock')(self.store_lock_path):
            document = self._load_trusted_fill_document()
            approval = document["approvals"].get(job_id)
            job = self._load_jobs_document()["jobs"].get(job_id)
            claim = self._load_coordinator_document()["claim"]
            if job is None or job.get("deletedAt") is not None or job["status"] != "in_progress":
                raise StoreError("trusted fill evaluation requires an in-progress job")
            if (
                claim is None or claim["jobId"] != job_id
                or self._now_datetime() >= self._parse_time(claim["expiresAt"])
            ):
                return {
                    "authorized": False,
                    "reasonCode": "claim_missing_or_expired",
                    "retryAllowed": False,
                    "attentionHandoff": False,
                }
            if approval is None:
                decision = {"authorized": False, "reasonCode": "approval_missing", "retryAllowed": False}
            elif approval["approvalRevision"] != observed.get("expectedApprovalRevision"):
                decision = {"authorized": False, "reasonCode": "approval_revision_mismatch", "retryAllowed": False}
            elif approval["claimId"] != claim["claimId"]:
                return {
                    "authorized": False,
                    "reasonCode": "claim_binding_mismatch",
                    "retryAllowed": False,
                    "attentionHandoff": False,
                }
            else:
                try:
                    current = self._trusted_fill_current_locked(
                        job_id, [item["answerRef"] for item in approval["answerBindings"]]
                    )
                except TrustedFillCurrentError as error:
                    decision = {
                        "authorized": False,
                        "reasonCode": error.reason_code,
                        "retryAllowed": False,
                    }
                else:
                    current["claimId"] = claim["claimId"]
                    try:
                        decision = _late('TRUSTED_FILL_MODULE').evaluate_approval(
                            approval, current, observed, self._now_datetime()
                        )
                    except _late('TRUSTED_FILL_MODULE').TrustedFillError as error:
                        raise StoreError(str(error)) from None
            if not decision["authorized"]:
                handed_off = self._trusted_fill_attention_handoff_locked(job, decision["reasonCode"])
                decision = {
                    **decision,
                    "attentionHandoff": True,
                    "job": {"id": handed_off["id"], "status": handed_off["status"], "revision": handed_off["revision"]},
                }
            else:
                decision["attentionHandoff"] = False
        return decision
